### PR TITLE
Add support for Series.operator.add in a new-style

### PR DESCRIPTION
**Pull request by kozlov-alexey**

### CODE REVIEW

File: sdc/datatypes/common_functions.py [hunk 1]
```diff
@@ -131,3 +132,19 @@ def _append_list_string_array_impl(A, B):
                 return new_data
 
             return _append_list_string_array_impl
+
+
+@numba.njit
+def _hpat_ensure_array_capacity(new_size, arr):
+    '''Function creating a copy of numpy array with a size more than specified'''
+    # TODO: replace this function with np.resize when supported by Numba
```
Comment (densmirn) [inline sdc/datatypes/common_functions.py]: `np.resize` resizes array in other way than this algorithm. The algorithm can generates array of size more than `new_size` and doesn't repeat values of `arr` in the resulting array. Looks like there is mismatch somewhere :)

File: sdc/datatypes/common_functions.py [hunk 2]
```diff
@@ -131,3 +132,19 @@ def _append_list_string_array_impl(A, B):
                 return new_data
 
             return _append_list_string_array_impl
+
+
+@numba.njit
+def _hpat_ensure_array_capacity(new_size, arr):
+    '''Function creating a copy of numpy array with a size more than specified'''
```
Comment (densmirn) [inline sdc/datatypes/common_functions.py]: Could you apply below recommendation everywhere in this PR?
[pep8 recommendation](https://www.python.org/dev/peps/pep-0008/#string-quotes): For triple-quoted strings, always use double quote characters to be consistent with the docstring convention in PEP 257.

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 3]
```diff
@@ -3646,3 +3649,343 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+    Returns
+    -------
+    :obj:`pandas.Series`
+        The result of the operation
+    """
+
+    _func_name = 'Operator add().'
+
+    if not isinstance(self, SeriesType):
+        raise TypingError('{} The object must be a pandas.series. Given: {}'.format(_func_name, self))
```
Comment (densmirn) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Could you use `TypeChecker` where it's possible?

File: sdc/datatypes/common_functions.py (hunk 1 above)
Comment (kozlov-alexey) [reply sdc/datatypes/common_functions.py]: Removed the comment at all)

File: sdc/datatypes/common_functions.py (hunk 2 above)
Comment (kozlov-alexey) [reply sdc/datatypes/common_functions.py]: Applied

File: sdc/datatypes/hpat_pandas_series_functions.py (hunk 3 above)
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Done

**Comment by kozlov-alexey** [comment]
@AlexanderKalistratov @Hardcode84 Can you please review too?

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 4]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+
+    # specializations for numeric series - TODO: support arithmetic operation on StringArrays
+    if (isinstance(other, types.Number)):
+        def hpat_pandas_series_add_scalar_impl(self, other):
+            return pandas.Series(self._data + other, self._index)
+
+        return hpat_pandas_series_add_scalar_impl
+
+    elif (isinstance(other, SeriesType)):
+        is_numeric_index = isinstance(self.index, (types.Array, types.NoneType))
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Shouldn't it be like: `(isinstance(self.index, types.NoneType) or isinstance(self.index, types.Array) and isinstance(self.index.dtype, types.Number)`. The same as check above? Or Check above should be the same as this one?

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 5]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+
+        if is_numeric_index:
+            ty_left_index_dtype = types.int64 if isinstance(self.index, types.NoneType) else self.index.dtype
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Looks like you can move this block to line 3958 (after line `if is_numeric_index == True:  # noqa`

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 6]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+            def hpat_pandas_series_add_impl(self, other):
+
+                if (len(self._data) == len(other._data)):
+                    return pandas.Series(numpy.asarray(self._data + other._data, numpy.float64))
+                else:
+                    min_data_size = min(len(self._data), len(other._data))
+                    max_data_size = max(len(self._data), len(other._data))
+                    new_data = numpy.empty(max_data_size, dtype=numpy.float64)
+                    new_data[:min_data_size] = self._data[:min_data_size] + other._data[:min_data_size]
+                    new_data[min_data_size:] = numpy.repeat(numpy.nan, max_data_size - min_data_size)
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_functions.py]: This behavior could be changed by parameters passed to `add` method of series.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.add.html#pandas.Series.add

This PR is about '+' operator, but what about 'add' method?

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 7]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+
+                    return pandas.Series(new_data, self._index)
+        else:
+            def hpat_pandas_series_add_impl(self, other):
+                left_index, right_index = self.index, other.index
+
+                # check if indexes are equal and series don't have to be aligned
+                if is_numeric_index == True:  # noqa
+                    if (numpy.array_equal(left_index, right_index)):
+                        return pandas.Series(numpy.asarray(self._data + other._data, numpy.float64),
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Are you sure about `numpy.float64`? shouldn't is be something like `common_dtype`?

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 8]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+                    # TODO: replace with StringArrays comparison
+                    is_index_equal = (len(self._index) == len(other._index)
+                                      and num_total_chars(self._index) == num_total_chars(other._index))
+                    for i in numpy.arange(len(self._index)):
+                        if (self._index[i] != other._index[i]
+                                or str_arr_is_na(self._index, i) is not str_arr_is_na(other._index, i)):
+                            is_index_equal = False
+
+                    if is_index_equal:
+                        return pandas.Series(numpy.asarray(self._data + other._data, numpy.float64),
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_functions.py]: Are you sure about `numpy.float64`? shouldn't is be something like `common_dtype`?

File: sdc/datatypes/hpat_pandas_series_functions.py (hunk 5 above)
Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Alright, most likely you can't. Since line 3958 inside jitted region

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 9]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+    ty_checker = TypeChecker('Operator add().')
+    ty_checker.check(self, SeriesType)
+
+    if not isinstance(other, (SeriesType, types.Number)):
+        ty_checker.raise_exc(other, 'pandas.series or scalar', 'other')
+
+    series_indexes_alignable = False
+    if isinstance(other, SeriesType):
+        if (other.index == string_array_type and self.index == string_array_type):
+            series_indexes_alignable = True
```
[the remark this answers is not in the document]
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Well, pandas will use dtype that can hold objects of diff types, i.e. resulting series will have index with dtype=object and all elements of resulting series will be NaN, since indexes won't align.

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 10]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+
+        if ((isinstance(self.index, types.NoneType) or
+             isinstance(self.index, types.Array) and isinstance(self.index.dtype, types.Number))
```
[the remark this answers is not in the document]
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Done

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 11]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+            and (isinstance(other.index, types.NoneType) or
+                 isinstance(other.index, types.Array) and isinstance(other.index.dtype, types.Number))):
```
[the remark this answers is not in the document]
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Added

File: sdc/datatypes/hpat_pandas_series_functions.py (hunk 4 above)
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Given that the above check passed, we can only check first series here, that's why it worked. But I agree that it's better to reuse the check from above, so I'll change to using 'none_or_numeric_indexes' variable in the next commit.

File: sdc/datatypes/hpat_pandas_series_functions.py (hunk 5 above)
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Well, yes, but this block can be moved to the else branch below, where it's probably a better place for it. Will do in the next commit.

File: sdc/datatypes/hpat_pandas_series_functions.py (hunk 6 above)
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: @AlexanderKalistratov Yes, I understand, I actually planned to have two different templates: one for binary operators and the other for add/sub/div/... methods, which can then be used with auto-generation script, because latter need special handling for method args and so on. But yes, I will address that in subsequent PRs.

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 12]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+
+@overload(hpat_join_series_indexes)
+def hpat_join_series_indexes_overload(left, right):
+    '''Function for joining arrays left and right in a way similar to pandas.join 'outer' algorithm'''
+
+    if (isinstance(left, types.Array) and isinstance(right, types.Array)):
+        np_dtypes = [numpy_support.as_dtype(left.dtype), numpy_support.as_dtype(right.dtype)]
+        np_common_dtype = numpy.find_common_type([], np_dtypes)
+        numba_common_dtype = numpy_support.from_dtype(np_common_dtype)
+        if (isinstance(left.dtype, types.Number) and isinstance(right.dtype, types.Number)):
```
[the remark this answers is not in the document]
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Fixed

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 13]
```diff
@@ -3646,3 +3649,344 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+                elif ridx[i] != -1:
+                    joined[i] = right[ridx[i]]
+                    if (str_arr_is_na(right, ridx[i])):
+                        str_arr_set_na(joined, i)
+                else:
+                    str_arr_set_na(joined, i)
+
+            return joined, lidx, ridx
+
+        return hpat_join_series_indexes_impl
```
[the remark this answers is not in the document]
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Yeap, done.

File: sdc/datatypes/hpat_pandas_series_functions.py (hunk 7 above)
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Yes, thank you, that's incorrect of course.

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_functions.py]: @kozlov-alexey I'm already not that sure it is incorrect. In case we are adding `nan` dtype should become numpy.float64 anyway

File: sdc/tests/test_series.py [hunk 14]
```diff
@@ -4392,6 +4615,7 @@ def test_series_pct_change_impl(S, periods=1, fill_method='pad', limit=None, fre
     "test_series_op4",
     "test_series_op6",
     "test_series_op7",
+    "test_series_operator_add_numeric_same_index_default",
```
Comment (PokhodenkoSA) [inline sdc/tests/test_series.py]: This code is not in master after #340.

File: sdc/datatypes/hpat_pandas_series_functions.py (hunk 7 above)
Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Right, just verified that the result of an operation in Pandas has unstable dtype: if indexes match it will be common numpy dtype of two series, but if one series has indexes absent in other, then np.nan will appear in data and resulting dtype will be float. So we have to use types.float64 due to type stability requirements.

File: sdc/tests/test_series.py (hunk 14 above)
Comment (kozlov-alexey) [reply sdc/tests/test_series.py]: Yes, thank you, I'll use new skip decorators.

File: sdc/datatypes/hpat_pandas_series_functions.py [hunk 15]
```diff
@@ -3795,3 +3798,132 @@ def hpat_pandas_series_pct_change_impl(self, periods=1, fill_method='pad', limit
...
+                if (len(self._data) == len(other._data)):
+                    return pandas.Series(numpy.asarray(self._data + other._data, numpy.float64))
+                else:
+                    min_data_size = min(len(self._data), len(other._data))
+                    max_data_size = max(len(self._data), len(other._data))
+                    new_data = numpy.empty(max_data_size, dtype=numpy.float64)
+                    new_data[:min_data_size] = self._data[:min_data_size] + other._data[:min_data_size]
+                    new_data[min_data_size:] = numpy.repeat(numpy.nan, max_data_size - min_data_size)
+
+                    return pandas.Series(new_data, self._index)
```
Comment (AlexanderKalistratov) [inline sdc/datatypes/hpat_pandas_series_functions.py]: @kozlov-alexey In some cases resulting series should have the same name as original one. This cases, seems to be are:
* series + scalar (i.e. A + 1)
* series + series, but both series has the same name

This is very important in order to implement DataFrame filters. I'm ok to do it in separate PR, but this must be done. Also tests for named series should be added

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_functions.py]: Well, after some more investigations, it's not important for filters, but still we should be consistent with pandas

Comment (AlexanderKalistratov) [reply sdc/datatypes/hpat_pandas_series_functions.py]: It looks like, at the moment we can't support second case (when two series have the same name) since it is runtime condition, but whether series has name or not we should know at compile time.

Comment (kozlov-alexey) [reply sdc/datatypes/hpat_pandas_series_functions.py]: @AlexanderKalistratov Thank you! I think we should address this in a separate PR. BTW, I just checked if there's any difference between '1 + A' and 'A + 1' and there's no, in both cases resulting series takes name from A. I'm going to dig in pandas sources to clarify this behavior.
